Display Statistics View per Student for their Submission Time

bivanalhar opened the pull request:
## Feature

We implemented Submission Time Due Table, which basically shows the submission status for each published, non-submitted assessments for any particular student. The objective is to give tutors or teachers the overview of which assessments have not yet been done by the students, and how far it is from the due date.

This feature is requested as part of the effort in knowing which student needs some extra help in finishing up each assessment, and also to give tutors time to warn the students to finish up their assignments, preferably before the due date.

## Overview

Since we want to focus on each student, the design, as well as the entry point, of this page is quite different from the other Course Statistics Page, as follows:
- Firstly, you need to go to the Course Statistics Students page
- Then, for the student that you might be interested in knowing their not-yet-submitted assessments, you can just click on their name, then the new tab will open, showing all those assessments

<img width="1528" alt="Screenshot 2024-04-18 at 5 37 25 PM" src="https://github.com/Coursemology/coursemology2/assets/16359075/e1b5073c-404c-479e-b716-e7da532a268d">

bivanalhar committed:
feat(assessment_statistics): provide page and API

- API merely sketch from FE (will complete implementation in BE)
- mock page for assessment statistics
- routes for assessments statistics fetching API
- at this point, only mock data is provided to FE from BE

bivanalhar committed:
feat(assessment_statistics): provide page and API

- API merely sketch from FE (will complete implementation in BE)
- mock page for assessment statistics
- routes for assessments statistics fetching API
- at this point, only mock data is provided to FE from BE

bivanalhar committed:
feat(assessment_stats): get all statistics info

- gather all info in BE, sending to FE
- code cleaning in BE

bivanalhar committed:
feat(assessment_stats): get all statistics info

- gather all info in BE, sending to FE
- code cleaning in BE

bivanalhar committed:
feat(assessment_statistics): display table

- using sortable and searchable table

bivanalhar committed:
fix(style): numLate before numAttempting

- also fix the styling in concern files
- default of avg and stdev to 0 if assessment_id not found in hash

bivanalhar committed:
refactor(stats): change in attempting

- include submitted also in counting attempting submissions
- use SQL to extract maximum grade

bivanalhar committed:
test(assessment): test for course stats assessment

bivanalhar committed:
feat(api): add API for course stats assessment

bivanalhar committed:
feat(api): add API for course stats assessment

bivanalhar committed:
fix(stats): remove statisticsTabs since unused

- already moved to index.tsx
- add Assessment Tab in index.tsx
- delete courseStatistics.js as already migrated to typescript

bivanalhar committed:
feat(stats): add tab for submission time info

- extend users API to include extract only students basic info
- prepare the Page for Submission Time (autocomplete field and space for table)

bivanalhar committed:
refactor(reference-time): include start_at

- needed for submission time table, to indicate the following
- if assessment has started respective to student's personal time

bivanalhar committed:
fix(tabs): default chosen tab should not be always 0

- it should match with the URL provided

bivanalhar committed:
feat(API): create API for fetching submission time

- depending on which student is chosen

bivanalhar committed:
feat(submission_time): get submission time info in BE

bivanalhar committed:
feat(translation): add Unreleased into status translations

bivanalhar committed:
feat(submission_time): add Submission Time Table

- fix rubocop issue

bivanalhar committed:
fix(submission): fix and add test

- course_user_id constrained only within current course
- add test for submission time

bivanalhar committed:
feat(timeline): pass isFixed and personalised time to FE

- to display properly and show user if the timeline is personalised or not

bivanalhar committed:
feat(submission_time): change route and page

- remove tab for submission time, entry through students stats table
- table design change to provide due date and due in
- include personalised indicator and real deadline
- change API routing for getting submission_time

bivanalhar committed:
fix(naming): submission_time -> submission_due

cysjonathan commented on app/controllers/concerns/course/statistics/reference_times_concern.rb:
```diff
@@ -17,28 +17,28 @@ def personal_end_at_hash(assessment_id_array, course_id)
 
       personal_times AS (
-        SELECT cu.id AS course_user_id, pt.end_at, pt.assessment_id
+        SELECT cu.id AS course_user_id, pt.start_at, pt.end_at, pt.fixed, pt.assessment_id
         FROM (
           SELECT course_users.id
           FROM course_users
           WHERE course_users.course_id = #{course_id}
         ) cu
         LEFT JOIN (
-          SELECT course_user_id, end_at, assessment_id
+          SELECT course_user_id, start_at, end_at, fixed, assessment_id
           FROM course_user_personal_end_at
         ) pt
         ON cu.id = pt.course_user_id
       ),
```
Why need nested subquery?
Can we do:
```sql
personal_times AS (
  SELECT cu.id AS course_user_id, pt.start_at, pt.end_at, pt.fixed, pt.assessment_id
  FROM course_users cu
  LEFT JOIN course_user_personal_end_at pt
  ON cu.id = pt.course_user_id
  WHERE cu.course_id = #{course_id}
)```

cysjonathan replied:
Same for the other queries

bivanalhar replied:
when making this nested subquery, my consideration is to make the code neater to read (having the code just like how u wrote it looks very condensed and not really that nice to read).

However, I think this is just about preference and I have adjusted the code to be like what you suggested above

cysjonathan commented on app/controllers/concerns/course/statistics/submissions_concern.rb:
```diff
@@ -89,8 +89,8 @@ def populate_hash_including_answers(student_hash, submissions)
       answers = answers_hash[submission.id]
-      end_at = @personal_end_at_hash[[@assessment.id, submitter_course_user.id]] ||
-               @reference_times_hash[@assessment.id]
+      _, end_at, = @personal_timeline_hash[[@assessment.id, submitter_course_user.id]] ||
+                   @reference_times_hash[@assessment.id]
```
Does this account for Multiple Reference Timelines (as in the timeline designer)?

bivanalhar replied:
yes, this functionality considers the personalised timeline to be the priority, then only using reference timeline if there's no personalised one

bivanalhar commented on app/helpers/course/statistics/aggregate_helper.rb:
```diff
@@ -1,14 +1,28 @@
     format("%2d #{t('time.day')} %02d:%02d:%02d", days, hours, minutes, seconds)
   end
+
+  def time_until_due(submitted_at, end_at)
```
I think `time_delta` is a very general phrase, even though what this function represents is actually how far the `submitted_at` is from the `end_at`.

The usage of `time_until_due` might be communicating the ideas better to the developer.

cysjonathan replied:
The point was that `time_until_due(submitted_at, end_at)` is behaving as a general function, time delta represents how far datetimestamp1 (submitted_at) is from datetimestamp2 (end_at).

`time_until_due` function signature sounds like `time_until_due(assessment_id, student_id)` which reports the time until the assessment is due for the given student (default time if no student is provided, or no personal times).

@ekowidianto your thoughts?

bivanalhar replied:
changed into `calculate_time_interval` and also the arguments naming is revised to `from` and `optional_to`

bivanalhar commented on app/views/course/statistics/aggregate/submission_due.json.jbuilder:
```diff
@@ -0,0 +1,48 @@
+    json.workflowState 'unstarted'
+  end
+
+  json.dueIn time_until_due(Time.now, end_at)
```
might be sufficient but I'm afraid the code's not communicative enough (right now it's just using `Time.now` since we are not considering the submitted/graded/published submissions).

But for the future, if we want to consider those kinds of submissions as well, the function `time_until_due` might be more general and can be used for that case as well.

cysjonathan commented on spec/controllers/course/statistics/aggregate_controller_spec.rb:
```diff
@@ -210,5 +210,77 @@
+          json_result = JSON.parse(response.body)
+          expect(json_result['assessments'].count).to eq(1)
+
+          expected_end_at = published_assessment.end_at.change(usec: published_assessment.end_at.usec / 1000 * 1000)
+          parsed_end_at = DateTime.parse(json_result['assessments'][0]['endAt']).
+                          change(usec: DateTime.parse(json_result['assessments'][0]['endAt']).
+                                       usec / 1000 * 1000)
+          parsed_reference_end_at = DateTime.parse(json_result['assessments'][0]['referenceEndAt']).
+                                    change(usec: DateTime.parse(json_result['assessments'][0]['referenceEndAt']).
+                                                 usec / 1000 * 1000)
+
+          expect(json_result['assessments'][0]['workflowState']).to eq('attempting')
+          expect(parsed_end_at).to eq(expected_end_at)
+          expect(parsed_reference_end_at).to eq(expected_end_at)
```
isn't there a `.to be_within(time_delta)` method that we can use instead of doing truncation like this?

bivanalhar replied:
you're right. I have done some more inspection and indeed, previously I was using this `be_within` function but I was doing it wrongly and thus it's unsuccessful

now that I have done it properly, it gets right now

bivanalhar commented on app/views/course/statistics/aggregate/submission_due.json.jbuilder:
```diff
@@ -0,0 +1,48 @@
+    json.submissionId submission_id
+    json.workflowState workflow_state
+  else
+    is_not_released_to_student = Time.now.to_i < start_at.to_i
```
we don't need that, you're right. I have deleted that accordingly

bivanalhar commented on app/views/course/statistics/aggregate/submission_due.json.jbuilder:
```diff
@@ -0,0 +1,48 @@
+json.name @all_students.first.name
+
+json.assessments @assessments do |assessment|
+  student_id = @all_students.first.id
```
oh ya sorry about this. this can be put also outside the do-loop, and might be saving the execution time a bit.

Even though the effect might not be significant, but I moved this out from the loop

bivanalhar commented on app/controllers/course/statistics/aggregate_controller.rb:
```diff
@@ -28,33 +30,65 @@ def all_students
   end
 
+  def submission_due
+    @all_students = [CourseUser.where(id: params[:student_id]).first]
```
this is because this `submission_due` API only accept one `student_id` from the frontend (the way this API is called is firstly to specify which student we would like to know more).

And the reason I'm using `@all_students` here is because my intention is to reuse the function `all_submissions_info`

bivanalhar commented on client/app/bundles/course/assessment/submission/constants.ts:
```diff
@@ -22,6 +22,13 @@ export const workflowStates = {
+};
+
+export const dueInStates = {
+  Safe: 'safe' as const,
+  AlmostLate: 'almostLate' as const,
+  Overdue: 'overdue' as const,
```
You got the point there, alright I'll change it.

bivanalhar commented on client/app/bundles/course/statistics/utils/dueTimeHelper.ts:
```diff
@@ -0,0 +1,53 @@
+
+  let dueStates = '';
+
+  if (days >= 3 && rawTime >= 0) {
```
got it. I've changed this

bivanalhar commented on app/controllers/course/statistics/aggregate_controller.rb:
```diff
@@ -28,33 +30,65 @@ def all_students
     fetch_all_assessment_related_statistics_hash
   end
 
+  def submission_due
```
I'm thinking that `submission_due` is also correct since we only expect 1 submission per assessment.
Nevertheless, I can change this into what you suggested as well

ekowidianto commented on app/controllers/course/statistics/aggregate_controller.rb:
```diff
@@ -2,9 +2,11 @@
 # This is named aggregate controller as naming this as course controller leads to name conflict issues
 class Course::Statistics::AggregateController < Course::Statistics::Controller
   before_action :preload_levels, only: [:all_students, :course_performance]
+  before_action :fetch_published_assessments, only: [:all_assessments, :submission_due]
```
```suggestion
  before_action :preload_assessments, only: [:all_assessments, :submission_due]
```
be consistent in the naming. see line 4 above.

bivanalhar replied:
got it.. though I will change it into `preload_published_assessments` since we actually only care about the published one, and ignoring the unpublished.

bivanalhar committed:
chore(naming): some minor changes in name

- simplify SQL script not to use redundant nesting code
- submission_due -> submissions_due
- fetch -> preload to make it consistent with other similar functions
- naming of dueInStates to be more consistent (notDue, almostDue, overDue)

bivanalhar committed:
refactor(submission-time): change UI

- student name in student stats link to user page
- add one column for clickable icon, link to overdue assessments
